core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -0,0 +1,469 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.WeightRampingUpStrategy.EndpointsInUpdatingEntry.EndpointAndStep;
+import com.linecorp.armeria.common.util.Ticker;
+
+final class WeightRampingUpStrategy implements EndpointSelectionStrategy {
```

Comment:
This class is pretty hard to understand from the code, I think it can use a high level design javadoc explaining what it does, how various stages interact with each other (select endpoint, ramp up weight, update list of endpoints). One thing I couldn't answer from looking at the code is whether when the endpoints list is updated, do we only ramp up new endpoints or ramp them all up again.

Also, did you consider handling ramping as part of "select endpoint" instead of using a background executor? I think it could be easier to reason about, not sure about effect on performance though since I guess it would require a lock, but that could stop once the endpoints are all ramped.

Reply:
> did you consider handling ramping as part of "select endpoint" instead of using a background executor?

I thought selecting endpoint and ramping together will make the code more complicated. If we handle it together, every time before selecting an endpoint, we should check whether there are endpoints that need to be ramped up. Then, we should ramp up the weight and select an endpoint using the new weights. 

> One thing I couldn't answer from looking at the code is whether when the endpoints list is updated, do we only ramp up new endpoints or ramp them all up again.

We only ramp up new endpoints. Should we ramp them all up?

> This class is pretty hard to understand from the code, I think it can use a high level design javadoc explaining what it does

Add more Javadocs and rename some variables. PTAL. 🙇